areas/mysql/src/AzureMcp.MySql/Services/MySqlService.cs

```diff
@@ -0,0 +1,482 @@
+
+        // Trim the result
+        cleanedQuery = cleanedQuery.Trim();
+
+        // Ensure the cleaned query is not empty
+        if (string.IsNullOrWhiteSpace(cleanedQuery))
+        {
+            throw new ArgumentException("Query cannot be empty after removing comments and whitespace.", nameof(query));
+        }
+
+        // Check for multiple SQL statements (semicolons that are not in comments)
+        var semicolonCount = cleanedQuery.Split(';').Length - 1;
+        if (semicolonCount > 1)
+        {
+            throw new InvalidOperationException("Multiple SQL statements are not allowed. Use only a single SELECT statement.");
+        }
+
+        // Regex pattern to detect common SQL injection techniques (comments are already removed)
+        var dangerPattern = new Regex(
+            @"(;|\b(union\s+select|drop\s+table|insert\s+into|update\s+|delete\s+from|or\s+1=1)\b)",
+            RegexOptions.IgnoreCase | RegexOptions.Compiled
+        );
+
+        // Check for common SQL injection patterns using regex
+        if (dangerPattern.IsMatch(cleanedQuery))
+        {
+            throw new InvalidOperationException("Query contains dangerous patterns that could indicate SQL injection attempts and is not allowed for security reasons.");
+        }
+
+        // List of dangerous SQL keywords that should be blocked
+        var queryUpper = cleanedQuery.ToUpperInvariant();
+
+        foreach (var keyword in DangerousKeywords)
+        {
+            if (queryUpper.Contains(keyword))
+            {
+                throw new InvalidOperationException($"Query contains dangerous keyword '{keyword}' which is not allowed for security reasons.");
+            }
+        }
+
+        // Check for character conversion functions that may be used for obfuscation
+        foreach (var func in ObfuscationFunctions)
+        {
+            if (queryUpper.Contains(func))
+            {
+                throw new InvalidOperationException($"Character conversion and obfuscation functions like '{func.TrimEnd('(')}' are not allowed for security reasons.");
+            }
+        }
+
+        // Additional validation: Only allow SELECT statements
+        var trimmedQuery = queryUpper.Trim();
+        var allowedStartPatterns = new[]
+        {
+            "SELECT"
+        };
+
+        bool isAllowed = allowedStartPatterns.Any(pattern => trimmedQuery.StartsWith(pattern));
+
+        if (!isAllowed)
+        {
+            throw new InvalidOperationException("Only SELECT statements are allowed for security reasons.");
+        }
+    }
+
+    public async Task<List<string>> ListDatabasesAsync(string subscriptionId, string resourceGroup, string user, string server)
+    {
+        try
+        {
+            var connectionString = await BuildConnectionStringAsync(server, user, "mysql");
+
+            await using var resource = await MySqlResource.CreateAsync(connectionString);
```

Comment:
Should we cache the instance?

Reply:
The expensive operation (getting Entra ID tokens) is already cached with proper expiration handling. Connection creation is relatively fast when using cached tokens.